src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -72,6 +79,204 @@ internal class Http2FrameWriter
         _scheduleInline = serviceContext.Scheduler == PipeScheduler.Inline;
 
         _hpackEncoder = new DynamicHPackEncoder(serviceContext.ServerOptions.AllowResponseHeaderCompression);
+
+        _channel = Channel.CreateBounded<Http2OutputProducer>(new BoundedChannelOptions(serviceContext.ServerOptions.Limits.Http2.MaxStreamsPerConnection)
+        {
+            AllowSynchronousContinuations = _scheduleInline
+        });
+
+        // This is null in tests sometimes
+        _window = connectionOutputFlowControl?.Available ?? 0;
+
+        _ = Task.Run(WriteToOutputPipe);
+    }
+
+    public void Schedule(Http2OutputProducer producer)
+    {
+        var scheduled = _channel.Writer.TryWrite(producer);
+
+        Debug.Assert(scheduled, $"Unable to schedule Stream {producer.Stream.StreamId}");
+    }
+
+    private async Task WriteToOutputPipe()
+    {
+        await foreach (var producer in _channel.Reader.ReadAllAsync())
```

Comment:
Is the only way this loop exits the writer being completed? I think that's currently missing.

Reply:
https://github.com/dotnet/aspnetcore/pull/40925/files#r837053258

Reply:
Fixed